src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,115 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_ARRAY_SIZE = 10;
+    private static final String OUT_OF_BOUNDS_MESSAGE = "Index out of bounds";
+    private static final String NO_SUCH_ELEMENT_MESSAGE = "No such element";
+    private int size;
+
+    private T[] elementData;
+
+    public ArrayList() {
+        this.elementData = (T[]) new Object[DEFAULT_ARRAY_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        elementDataGrow(size);
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(OUT_OF_BOUNDS_MESSAGE);
+        } else {
+            System.arraycopy(elementData, index,
+                    elementData, index + 1,
+                    size - index);
+            elementData[index] = value;
+            size++;
+            elementDataGrow(size);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
+        checkIndex(index);
+        elementData[index] = value;
 
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedElement;
+        if ((size == elementData.length) && index + 1 == size) {
+            size--;
+            removedElement = elementData[index];
+            return removedElement;
+        }
+        removedElement = elementData[index];
+        System.arraycopy(elementData, index + 1, elementData, index, size - index);
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (elementData[i] == null && element == null) {
+                size--;
+                return null;
+            }
+            if (elementData[i] != null && elementData[i].equals(element)) {
+                return remove(i);
+            }
```

Comment:
if (array[i] ==  element ||  element != null && element.equals(array[i])) {
                return remove(i);
            }
you can simplify it